Provide consistent examples of state management for all SDKs (#2000)
dapr/docs: daprdocs/content/en/developing-applications/building-blocks/state-management/howto-get-save-state.md

- [Dapr CLI]({{< ref install-dapr-cli.md >}})
- Initialized [Dapr environment]({{< ref install-dapr-selfhost.md >}})

## Example:

> The below code examples loosely describe an application that processes orders. In the examples, there is one service - an order processing service. This service has Dapr sidecar and the order processing service uses Dapr to store state in redis state store.

Reviewer: ```suggestion
The below code examples loosely describe an application that processes orders. In the examples, there is an order processing service which has a Dapr sidecar. The order processing service uses Dapr to store state in a Redis state store.
```

> <img src="/images/state_management_eg.png" width=1000 alt="Diagram showing state management of example service">

Reviewer 2: @amulyavarote this image does not load correctly when running locally.

Reviewer 2: Also, can we rename the image from "state_management_eg" to "building-block-state-management-example"

Can we please follow this convention for all the other images too across building blocks?

```csharp

//dependencies
using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Threading.Tasks;
using Dapr.Client;
using Microsoft.AspNetCore.Mvc;
using System.Threading;
using System.Text.Json;

//code
namespace EventService
{
    class Program
    {
        static async Task Main(string[] args)
        {
            string DAPR_STORE_NAME = "statestore";

            //Calling service multiple times with 5 seconds gap in between the calls
            while(true) {
                System.Threading.Thread.Sleep(5000);
                Random random = new Random();
                int orderId = random.Next(1,1000);

                //Using Dapr SDK to save and get state
                using var client = new DaprClientBuilder().Build();
                await client.SaveStateAsync(DAPR_STORE_NAME, "order_1", orderId.ToString());
                await client.SaveStateAsync(DAPR_STORE_NAME, "order_2", orderId.ToString());
                var result = await client.GetStateAsync<string>(DAPR_STORE_NAME, orderId.ToString());
                Console.WriteLine("Result after get: " + result);
            }
        }
    }
}

```

> Once saved run the following command to launch a Dapr sidecar and run the application:

Reviewer: I think the "once saved" part may get confusing here, suggest something like:

```suggestion
Navigate to the directory containing the above code, then run the following command to launch a Dapr sidecar and run the application:
```

import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import reactor.core.publisher.Mono;
import java.util.Random;
import java.util.concurrent.TimeUnit;

//code
@SpringBootApplication
public class OrderProcessingServiceApplication {

	private static final Logger log = LoggerFactory.getLogger(OrderProcessingServiceApplication.class);

	public static void main(String[] args) throws InterruptedException{
        String STATE_STORE_NAME = "statestore";

        //Calling service multiple times with 5 seconds gap in between the calls
		while(true) {
			TimeUnit.MILLISECONDS.sleep(5000);
			Random random = new Random();
			int orderId = random.nextInt(1000-1) + 1;

            //Using Dapr SDK to save and get state
			DaprClient client = new DaprClientBuilder().build();
			client.saveState(STATE_STORE_NAME, "order_1", Integer.toString(orderId)).block();
			client.saveState(STATE_STORE_NAME, "order_2", Integer.toString(orderId)).block();
			Mono<State<String>> result = client.getState(STATE_STORE_NAME, "order_1", String.class);
			log.info("Result after get" + result);
		}
	}
}

```

> Once saved run the following command to launch a Dapr sidecar and run the application:

Reviewer: I think the "once saved" part may get confusing here, suggest something like:

```suggestion
Navigate to the directory containing the above code, then run the following command to launch a Dapr sidecar and run the application:
```

```bash
dapr run --app-id orderprocessing --app-port 6001 --dapr-http-port 3601 --dapr-grpc-port 60001 mvn spring-boot:run
```

{{% /codetab %}}

{{% codetab %}}

```python

#dependencies
import random
from time import sleep    
import requests
import logging
from dapr.clients import DaprClient
from dapr.clients.grpc._state import StateItem
from dapr.clients.grpc._request import TransactionalStateOperation, TransactionOperationType

#code
logging.basicConfig(level = logging.INFO)
    
DAPR_STORE_NAME = "statestore"

#Calling service multiple times with 5 seconds gap in between the calls
while True:
    sleep(random.randrange(50, 5000) / 1000)
    orderId = random.randint(1, 1000)

    #Using Dapr SDK to save and get state
    with DaprClient() as client:
        client.save_state(DAPR_STORE_NAME, "order_1", str(orderId)) 
        result = client.get_state(DAPR_STORE_NAME, "order_1")
        logging.info('Result after get: ' + str(result))

```

> Once saved run the following command to launch a Dapr sidecar and run the application:

Reviewer: I think the "once saved" part may get confusing here, suggest something like:

```suggestion
Navigate to the directory containing the above code, then run the following command to launch a Dapr sidecar and run the application:
```

```bash
dapr run --app-id orderprocessing --app-port 6001 --dapr-http-port 3601 --dapr-grpc-port 60001 -- python3 OrderProcessingService.py
```

{{% /codetab %}}

{{% codetab %}}

```go

//dependencies
import (
	"context"
	"log"
	"math/rand"
	"time"
	"strconv"
	dapr "github.com/dapr/go-sdk/client"

)

//code
func main() {

    STATE_STORE_NAME := "statestore"

    //Calling service multiple times with 5 seconds gap in between the calls
	for i := 0; i < 10; i++ {
		time.Sleep(5000)
		orderId := rand.Intn(1000-1) + 1

        //Using Dapr SDK to save and get state
		client, err := dapr.NewClient()
		if err != nil {
			panic(err)
		}
		defer client.Close()
		ctx := context.Background()

		if err := client.SaveState(ctx, STATE_STORE_NAME, "order_1", []byte(strconv.Itoa(orderId))); err != nil {
			panic(err)
		}
		
		result, err := client.GetState(ctx, STATE_STORE_NAME, "order_1")
		if err != nil {
			panic(err)
		}

		log.Println("Result after get: ")
		log.Println(result)
	}
}

```

> Once saved run the following command to launch a Dapr sidecar and run the application:

Reviewer: I think the "once saved" part may get confusing here, suggest something like:

```suggestion
Navigate to the directory containing the above code, then run the following command to launch a Dapr sidecar and run the application:
```

```

{{% /codetab %}}

{{% codetab %}}

```javascript

//dependencies
import { DaprClient, HttpMethod, CommunicationProtocolEnum } from 'dapr-client'; 

//code
const daprHost = "127.0.0.1"; 

var main = function() {
    const STATE_STORE_NAME = "statestore";

    //Calling service multiple times with 5 seconds gap in between the calls
    for(var i=0;i<10;i++) {
        sleep(5000);
        var orderId = Math.floor(Math.random() * (1000 - 1) + 1);

        //Using Dapr SDK to save and get state
        const client = new DaprClient(daprHost, process.env.DAPR_HTTP_PORT, CommunicationProtocolEnum.HTTP);
        await client.state.save(STATE_STORE_NAME, [
            {
                key: "order_1",
                value: orderId.toString()
            },
            {
                key: "order_2",
                value: orderId.toString()
            }
        ]);
        var result = await client.state.get(STATE_STORE_NAME, "order_1");
        console.log("Result after get: " + result);
    }
}

function sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}

main();

```

> Once saved run the following command to launch a Dapr sidecar and run the application:

Reviewer: I think the "once saved" part may get confusing here, suggest something like:

```suggestion
Navigate to the directory containing the above code, then run the following command to launch a Dapr sidecar and run the application:
```

```bash
dapr run --app-id orderprocessing --app-port 6001 --dapr-http-port 3601 --dapr-grpc-port 60001 npm start
```

{{% /codetab %}}

{{% codetab %}}
Begin by launching a Dapr sidecar:

```bash
dapr run --app-id orderprocessing --dapr-http-port 3601
```

Then in a separate terminal save a key/value pair into your statestore:
```bash
curl -X POST -H "Content-Type: application/json" -d '[{ "key": "order_1", "value": "250"}]' http://localhost:3601/v1.0/state/statestore
```

Now get the state you just saved:
```bash
curl http://localhost:3601/v1.0/state/statestore/order_1
```

> You can also restart your sidecar and try retrieving state again to see that state persists separate from the app.

Reviewer: ```suggestion
Restart your sidecar and try retrieving state again to observe that state persists separately from the app.
```

{{% /codetab %}}

{{% codetab %}}

Begin by launching a Dapr sidecar:

```bash
dapr --app-id orderprocessing --port 3601 run
```

Then in a separate terminal save a key/value pair into your statestore:
```powershell
Invoke-RestMethod -Method Post -ContentType 'application/json' -Body '[{"key": "order_1", "value": "250"}]' -Uri 'http://localhost:3601/v1.0/state/statestore'
```

Now get the state you just saved:
```powershell
Invoke-RestMethod -Uri 'http://localhost:3601/v1.0/state/statestore/order_1'
```

> You can also restart your sidecar and try retrieving state again to see that state persists separate from the app.

Reviewer: ```suggestion
Restart your sidecar and try retrieving state again to observe that state persists separately from the app.
```

{{% /codetab %}}

{{< /tabs >}}

## Step 3: Delete state

Below are code examples that leverage Dapr SDKs for deleting the state.

{{< tabs Dotnet Java Python Go Javascript "HTTP API (Bash)" "HTTP API (PowerShell)">}}

{{% codetab %}}

```csharp

//dependencies
using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Threading.Tasks;
using Dapr.Client;
using Microsoft.AspNetCore.Mvc;
using System.Threading;
using System.Text.Json;

//code
namespace EventService
{
    class Program
    {
        static async Task Main(string[] args)
        {
            string DAPR_STORE_NAME = "statestore";
            //Using Dapr SDK to delete the state
            using var client = new DaprClientBuilder().Build();
            await client.DeleteStateAsync(DAPR_STORE_NAME, "order_1", cancellationToken: cancellationToken);
        }
    }
}

```

> Once saved run the following command to launch a Dapr sidecar and run the application:

Reviewer: I think the "once saved" part may get confusing here, suggest something like:

```suggestion
Navigate to the directory containing the above code, then run the following command to launch a Dapr sidecar and run the application:
```

```bash
dapr run --app-id orderprocessing --app-port 6001 --dapr-http-port 3601 --dapr-grpc-port 60001 dotnet run
```

{{% /codetab %}}

{{% codetab %}}

```java

//dependencies
import io.dapr.client.DaprClient;
import io.dapr.client.DaprClientBuilder;
import io.dapr.client.domain.State;
import io.dapr.client.domain.TransactionalStateOperation;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import java.util.Random;
import java.util.concurrent.TimeUnit;

//code
@SpringBootApplication
public class OrderProcessingServiceApplication {

	private static final Logger log = LoggerFactory.getLogger(OrderProcessingServiceApplication.class);

	public static void main(String[] args) throws InterruptedException{
        String STATE_STORE_NAME = "statestore";

        //Using Dapr SDK to delete the state
        DaprClient client = new DaprClientBuilder().build();
        String storedEtag = client.getState(STATE_STORE_NAME, "order_1", String.class).block().getEtag();
        client.deleteState(STATE_STORE_NAME, "order_1", storedEtag, null).block();
	}
}

```

> Once saved run the following command to launch a Dapr sidecar and run the application:

Reviewer: I think the "once saved" part may get confusing here, suggest something like:

```suggestion
Navigate to the directory containing the above code, then run the following command to launch a Dapr sidecar and run the application:
```

```bash
dapr run --app-id orderprocessing --app-port 6001 --dapr-http-port 3601 --dapr-grpc-port 60001 mvn spring-boot:run
```

{{% /codetab %}}

{{% codetab %}}

```python

#dependencies
import random
from time import sleep    
import requests
import logging
from dapr.clients import DaprClient
from dapr.clients.grpc._state import StateItem
from dapr.clients.grpc._request import TransactionalStateOperation, TransactionOperationType

#code
logging.basicConfig(level = logging.INFO)
    
DAPR_STORE_NAME = "statestore"

#Using Dapr SDK to delete the state
with DaprClient() as client:
    client.delete_state(store_name=DAPR_STORE_NAME, key="order_1")

```

> Once saved run the following command to launch a Dapr sidecar and run the application:

Reviewer: I think the "once saved" part may get confusing here, suggest something like:

```suggestion
Navigate to the directory containing the above code, then run the following command to launch a Dapr sidecar and run the application:
```

```bash
dapr run --app-id orderprocessing --app-port 6001 --dapr-http-port 3601 --dapr-grpc-port 60001 -- python3 OrderProcessingService.py
```

{{% /codetab %}}

{{% codetab %}}

```go

//dependencies
import (
	"context"
	dapr "github.com/dapr/go-sdk/client"

)

//code
func main() {

    STATE_STORE_NAME := "statestore"

    //Using Dapr SDK to delete the state
    client, err := dapr.NewClient()
    if err != nil {
        panic(err)
    }
    defer client.Close()
    ctx := context.Background()

    if err := client.DeleteState(ctx, STATE_STORE_NAME, "order_1"); err != nil {
        panic(err)
    }
}

```

> Once saved run the following command to launch a Dapr sidecar and run the application:

Reviewer: I think the "once saved" part may get confusing here, suggest something like:

```suggestion
Navigate to the directory containing the above code, then run the following command to launch a Dapr sidecar and run the application:
```

```bash
dapr run --app-id orderprocessing --app-port 6001 --dapr-http-port 3601 --dapr-grpc-port 60001 go run OrderProcessingService.go
```

{{% /codetab %}}

{{% codetab %}}

```javascript

//dependencies
import { DaprClient, HttpMethod, CommunicationProtocolEnum } from 'dapr-client'; 

//code
const daprHost = "127.0.0.1"; 

var main = function() {
    const STATE_STORE_NAME = "statestore";

    //Using Dapr SDK to save and get state
    const client = new DaprClient(daprHost, process.env.DAPR_HTTP_PORT, CommunicationProtocolEnum.HTTP);
    await client.state.delete(STATE_STORE_NAME, "order_1"); 
}

function sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}

main();

```

> Once saved run the following command to launch a Dapr sidecar and run the application:

Reviewer: I think the "once saved" part may get confusing here, suggest something like:

```suggestion
Navigate to the directory containing the above code, then run the following command to launch a Dapr sidecar and run the application:
```

```bash
dapr run --app-id orderprocessing --app-port 6001 --dapr-http-port 3601 --dapr-grpc-port 60001 npm start
```

{{% /codetab %}}

{{% codetab %}}

> With the same dapr instance running from above run:

Reviewer: ```suggestion
With the same Dapr instance running from above run:
```

```bash
curl -X DELETE 'http://localhost:3601/v1.0/state/statestore/order_1'
```
Try getting state again and note that no value is returned.
{{% /codetab %}}

{{% codetab %}}

> With the same dapr instance running from above run:

Reviewer: ```suggestion
With the same Dapr instance running from above run:
```

```powershell
Invoke-RestMethod -Method Delete -Uri 'http://localhost:3601/v1.0/state/statestore/order_1'
```
Try getting state again and note that no value is returned.
{{% /codetab %}}

{{< /tabs >}}

## Step 4: Save and retrieve multiple states

Below are code examples that leverage Dapr SDKs for saving and retrieving multiple states.

> {{< tabs Java Python Javascript "HTTP API (Bash)" "HTTP API (PowerShell)">}}

Reviewer: are .NET and Go SDK examples excluded on purpose?

```java

//dependencies
import io.dapr.client.DaprClient;
import io.dapr.client.DaprClientBuilder;
import io.dapr.client.domain.State;
import io.dapr.client.domain.TransactionalStateOperation;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import reactor.core.publisher.Mono;
import java.util.Random;
import java.util.concurrent.TimeUnit;

//code
@SpringBootApplication
public class OrderProcessingServiceApplication {

	private static final Logger log = LoggerFactory.getLogger(OrderProcessingServiceApplication.class);

	public static void main(String[] args) throws InterruptedException{
        String STATE_STORE_NAME = "statestore";

        //Calling service multiple times with 5 seconds gap in between the calls
		while(true) {
			TimeUnit.MILLISECONDS.sleep(5000);
			Random random = new Random();
			int orderId = random.nextInt(1000-1) + 1;

            //Using Dapr SDK to retrieve multiple states
			DaprClient client = new DaprClientBuilder().build();
			Mono<List<State<String>>> resultBulk = client.getBulkState(STATE_STORE_NAME,
					Arrays.asList("order_1", "order_2"), String.class);
		}
	}
}

```

> Once saved run the following command to launch a Dapr sidecar and run the application:

Reviewer: I think the "once saved" part may get confusing here, suggest something like:

```suggestion
Navigate to the directory containing the above code, then run the following command to launch a Dapr sidecar and run the application:
```

```bash
dapr run --app-id orderprocessing --app-port 6001 --dapr-http-port 3601 --dapr-grpc-port 60001 mvn spring-boot:run
```

{{% /codetab %}}

{{% codetab %}}

```python

#dependencies
import random
from time import sleep    
import requests
import logging
from dapr.clients import DaprClient
from dapr.clients.grpc._state import StateItem
from dapr.clients.grpc._request import TransactionalStateOperation, TransactionOperationType

#code
logging.basicConfig(level = logging.INFO)
    
DAPR_STORE_NAME = "statestore"

#Calling service multiple times with 5 seconds gap in between the calls
while True:
    sleep(random.randrange(50, 5000) / 1000)
    orderId = random.randint(1, 1000)

    #Using Dapr SDK to save and retrieve multiple states
    with DaprClient() as client:
        client.save_bulk_state(store_name=DAPR_STORE_NAME, states=[StateItem(key="order_2", value=str(orderId))])
        result = client.get_bulk_state(store_name=DAPR_STORE_NAME, keys=["order_1", "order_2"], states_metadata={"metakey": "metavalue"}).items
        logging.info('Result after get bulk: ' + str(result))

```

> Once saved run the following command to launch a Dapr sidecar and run the application:

Reviewer: I think the "once saved" part may get confusing here, suggest something like:

```suggestion
Navigate to the directory containing the above code, then run the following command to launch a Dapr sidecar and run the application:
```

```

{{% /codetab %}}

{{% codetab %}}

```javascript

//dependencies
import { DaprClient, HttpMethod, CommunicationProtocolEnum } from 'dapr-client'; 

//code
const daprHost = "127.0.0.1"; 

var main = function() {
    const STATE_STORE_NAME = "statestore";

    //Calling service multiple times with 5 seconds gap in between the calls
    for(var i=0;i<10;i++) {
        sleep(5000);
        var orderId = Math.floor(Math.random() * (1000 - 1) + 1);

        //Using Dapr SDK to save and retrieve multiple states
        const client = new DaprClient(daprHost, process.env.DAPR_HTTP_PORT, CommunicationProtocolEnum.HTTP);
        await client.state.save(STATE_STORE_NAME, [
            {
                key: "order_1",
                value: orderId.toString()
            },
            {
                key: "order_2",
                value: orderId.toString()
            }
        ]);
        result = await client.state.getBulk(STATE_STORE_NAME, ["order_1", "order_2"]);
        console.log("Result after get bulk: " + result);
    }
}

function sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}

main();

```

> Once saved run the following command to launch a Dapr sidecar and run the application:

Reviewer: I think the "once saved" part may get confusing here, suggest something like:

```suggestion
Navigate to the directory containing the above code, then run the following command to launch a Dapr sidecar and run the application:
```

```bash
dapr run --app-id orderprocessing --app-port 6001 --dapr-http-port 3601 --dapr-grpc-port 60001 npm start
```

{{% /codetab %}}

{{% codetab %}}

> With the same dapr instance running from above save two key/value pairs into your statestore:

Reviewer: ```suggestion
With the same Dapr instance running from above save two key/value pairs into your statestore:
```

```bash
curl -X POST -H "Content-Type: application/json" -d '[{ "key": "order_1", "value": "250"}, { "key": "order_2", "value": "550"}]' http://localhost:3601/v1.0/state/statestore
```

Now get the states you just saved:
```bash
curl -X POST -H "Content-Type: application/json" -d '{"keys":["order_1", "order_2"]}' http://localhost:3601/v1.0/state/statestore/bulk
```
{{% /codetab %}}

{{% codetab %}}

> With the same dapr instance running from above save two key/value pairs into your statestore:

Reviewer: ```suggestion
With the same Dapr instance running from above save two key/value pairs into your statestore:
```

State transactions require a state store that supports multi-item transactions. Visit the [supported state stores page]({{< ref supported-state-stores >}}) page for a full list. Note that the default Redis container created in a self-hosted environment supports them.
{{% /alert %}}

Below are code examples that leverage Dapr SDKs for performing state transactions.

> {{< tabs Dotnet Java Python Javascript "HTTP API (Bash)" "HTTP API (PowerShell)">}}

Reviewer: Is the Go SDK example excluded on purpose?

Author: Yes. No supporting sdk method available.

{
    class Program
    {
        static async Task Main(string[] args)
        {
            string DAPR_STORE_NAME = "statestore";
            //Calling service multiple times with 5 seconds gap in between the calls
            while(true) {
                System.Threading.Thread.Sleep(5000);
                Random random = new Random();
                int orderId = random.Next(1,1000);

                //Using Dapr SDK to perform the state transactions
                using var client = new DaprClientBuilder().Build();
                var requests = new List<StateTransactionRequest>()
                {
                    new StateTransactionRequest("order_3", JsonSerializer.SerializeToUtf8Bytes(orderId.ToString()), StateOperationType.Upsert),
                    new StateTransactionRequest("order_2", null, StateOperationType.Delete)
                };
                CancellationTokenSource source = new CancellationTokenSource();
                CancellationToken cancellationToken = source.Token;
                await client.ExecuteStateTransactionAsync(DAPR_STORE_NAME, requests, cancellationToken: cancellationToken);
            }
        }
    }
}

```

> Once saved run the following command to launch a Dapr sidecar and run the application:

Reviewer: I think the "once saved" part may get confusing here, suggest something like:

```suggestion
Navigate to the directory containing the above code, then run the following command to launch a Dapr sidecar and run the application:
```

import java.util.Random;
import java.util.concurrent.TimeUnit;

//code
@SpringBootApplication
public class OrderProcessingServiceApplication {

	private static final Logger log = LoggerFactory.getLogger(OrderProcessingServiceApplication.class);

	public static void main(String[] args) throws InterruptedException{
        String STATE_STORE_NAME = "statestore";

        //Calling service multiple times with 5 seconds gap in between the calls
		while(true) {
			TimeUnit.MILLISECONDS.sleep(5000);
			Random random = new Random();
			int orderId = random.nextInt(1000-1) + 1;

            //Using Dapr SDK to perform the state transactions
			DaprClient client = new DaprClientBuilder().build();
			List<TransactionalStateOperation<?>> operationList = new ArrayList<>();
			operationList.add(new TransactionalStateOperation<>(TransactionalStateOperation.OperationType.UPSERT,
					new State<>("order_3", Integer.toString(orderId), "")));
			operationList.add(new TransactionalStateOperation<>(TransactionalStateOperation.OperationType.DELETE,
					new State<>("order_2")));
			client.executeStateTransaction(STATE_STORE_NAME, operationList).block();
		}
	}
}

```

> Once saved run the following command to launch a Dapr sidecar and run the application:

Reviewer: I think the "once saved" part may get confusing here, suggest something like:

```suggestion
Navigate to the directory containing the above code, then run the following command to launch a Dapr sidecar and run the application:
```

#dependencies
import random
from time import sleep    
import requests
import logging
from dapr.clients import DaprClient
from dapr.clients.grpc._state import StateItem
from dapr.clients.grpc._request import TransactionalStateOperation, TransactionOperationType

#code
logging.basicConfig(level = logging.INFO)
    
DAPR_STORE_NAME = "statestore"

#Calling service multiple times with 5 seconds gap in between the calls
while True:
    sleep(random.randrange(50, 5000) / 1000)
    orderId = random.randint(1, 1000)

    #Using Dapr SDK to perform the state transactions
    with DaprClient() as client:
        client.execute_state_transaction(store_name=DAPR_STORE_NAME, operations=[
            TransactionalStateOperation(
                operation_type=TransactionOperationType.upsert,
                key="order_3",
                data=str(orderId)),
            TransactionalStateOperation(key="order_3", data=str(orderId)),
            TransactionalStateOperation(
                operation_type=TransactionOperationType.delete,
                key="order_2",
                data=str(orderId)),
            TransactionalStateOperation(key="order_2", data=str(orderId))
        ])

```

> Once saved run the following command to launch a Dapr sidecar and run the application:

Reviewer: I think the "once saved" part may get confusing here, suggest something like:

```suggestion
Navigate to the directory containing the above code, then run the following command to launch a Dapr sidecar and run the application:
```

```

{{% /codetab %}}

{{% codetab %}}

```javascript

//dependencies
import { DaprClient, HttpMethod, CommunicationProtocolEnum } from 'dapr-client'; 

//code
const daprHost = "127.0.0.1"; 

var main = function() {
    const STATE_STORE_NAME = "statestore";

    //Calling service multiple times with 5 seconds gap in between the calls
    for(var i=0;i<10;i++) {
        sleep(5000);
        var orderId = Math.floor(Math.random() * (1000 - 1) + 1);

        //Using Dapr SDK to save and retrieve multiple states
        const client = new DaprClient(daprHost, process.env.DAPR_HTTP_PORT, CommunicationProtocolEnum.HTTP);
        await client.state.transaction(STATE_STORE_NAME, [
            {
                operation: "upsert",
                request: {
                    key: "order_3",
                    value: orderId.toString()
                }
            },
            {
                operation: "delete",
                request: {
                    key: "order_2"
                }
            }
        ]);
    }
}

function sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}

main();

```

> Once saved run the following command to launch a Dapr sidecar and run the application:

Reviewer: I think the "once saved" part may get confusing here, suggest something like:

```suggestion
Navigate to the directory containing the above code, then run the following command to launch a Dapr sidecar and run the application:
```

```bash
dapr run --app-id orderprocessing --app-port 6001 --dapr-http-port 3601 --dapr-grpc-port 60001 npm start
```

{{% /codetab %}}

{{% codetab %}}

> With the same dapr instance running from above perform two state transactions:

Reviewer: ```suggestion
With the same Dapr instance running from above perform two state transactions:
```

```bash
curl -X POST -H "Content-Type: application/json" -d '{"operations": [{"operation":"upsert", "request": {"key": "order_1", "value": "250"}}, {"operation":"delete", "request": {"key": "order_2"}}]}' http://localhost:3601/v1.0/state/statestore/transaction
```

Now see the results of your state transactions:
```bash
curl -X POST -H "Content-Type: application/json" -d '{"keys":["order_1", "order_2"]}' http://localhost:3601/v1.0/state/statestore/bulk
```
{{% /codetab %}}

{{% codetab %}}

> With the same dapr instance running from above save two key/value pairs into your statestore:

Reviewer: ```suggestion
With the same Dapr instance running from above save two key/value pairs into your statestore:
```
